### ISSUE TITLE
Add linear algebra operations and their derivatives

### ISSUE THREAD

**Issue by dan-zheng** [tensorflow/swift-apis#562]
Swift for TensorFlow should provide standard linear algebra operations from [`tf.linalg`](https://www.tensorflow.org/api_docs/python/tf/linalg)/[`np.linalg`](https://docs.scipy.org/doc/numpy-1.13.0/reference/routines.linalg.html).

Some of these operations are already accessible via [TensorFlow bindings](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Bindings/RawOpsGenerated.swift): `_Raw.cholesky`, `_Raw.choleskyGrad`.

Reference TensorFlow derivative implementations: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/linalg_grad.py

Mirror of [TF-980](https://bugs.swift.org/browse/TF-980).

**Comment by awav** [comment]
This is not a full list of operations, but it would be good to start with this:

- [ ] Norm
- [ ] Trace
- [ ] Cholesky
- [ ] Triangular Solve
- [ ] Cholesky Solve
- [ ] Diag (make a [batch] tensor from [batch] diagonal)
- [ ] Diag Part (extract [batch] diagonal)
- [ ] Einsum
- [ ] Eye
- [ ] Logdet
- [ ] Eigh

**Comment by awav** [comment]
@dan-zheng , how can I add a new binding [here](https://github.com/tensorflow/swift-apis/blob/master/Sources/TensorFlow/Bindings/RawOpsGenerated.swift)?

**Comment by Shashi456** [comment]
@awav you can't add new bindings in that file, You can check where how its generated [here](https://github.com/tensorflow/swift-apis/tree/master/Sources/TensorFlow/Bindings).

**Comment by awav** [comment]
@Shashi456 , yeah, I've checked it. But, it is not clear why for e.g. "TriangularSolve" is not reflected in the [RawOpsGenerated.swift](https://raw.githubusercontent.com/tensorflow/swift-apis/master/Sources/TensorFlow/Bindings/RawOpsGenerated.swift) file, but "TriangularSolve" is in the [ops.pbtxt](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/ops.pbtxt). Does S4TF uses different protobuf file?

**Comment by jon-tow** [comment]
Hey @awav. I believe `triangular_solve` is aliased under `matrixTriangularSolve`.

**Comment by dan-zheng** [comment]
> @Shashi456 , yeah, I've checked it. But, it is not clear why for e.g. "TriangularSolve" is not reflected in the RawOpsGenerated.swift file, but "TriangularSolve" is in the ops.pbtxt. Does S4TF uses different protobuf file?

Swift for TensorFlow uses the same `ops.pbtxt` - [here's the script](https://github.com/tensorflow/swift-apis/tree/master/Sources/TensorFlow/Bindings) that generates op bindings.

`TensorFlow.Raw` and `ops.pbtxt` are consistent, both define `matrixTriangularSolve` and `batchMatrixTriangularSolve`:

![image](https://user-images.githubusercontent.com/5590046/69490985-deb19880-0e43-11ea-9943-da0eb83fe741.png)

**Comment by Shashi456** [comment]
@awav are you working on supporting these ops? could i ask what ops you are working on currently?

**Comment by awav** [comment]
@Shashi456, solvers and trace for now

@dan-zheng, you can mark `diag` and `diag_part` as finished. They were implemented in https://github.com/tensorflow/swift-apis/pull/571

**Comment by awav** [comment]
@Shashi456, @dan-zheng hello guys, sorry for the delay with PRs for solvers and other promised ops. I'm at NeurIPS and don't have much time to finish the work. Thanks!

**Comment by awav** [comment]
@dan-zheng, `trace` can be marked now as done.

**Comment by SumanSudhir** [comment]
@Shashi456 do swift for TensorFlow require linear algebra operator like [tf.expm](https://www.tensorflow.org/api_docs/python/tf/linalg/expm) and [tf.logm](https://www.tensorflow.org/api_docs/python/tf/linalg/logm) . It is not mentioned in above list but I am willing to work on it.

**Comment by pedronahum** [comment]
Hi, 
If I can add to the wishlist, the following matrix operations could be useful:
- QR eigen decomposition with explicit shift (Wilkinson shif)
- the Schur decomposition
